core/state/database.go

```diff
@@ -607,10 +624,37 @@ func (db *cachingDB) LoadTransitionState(root common.Hash) {
 		}
 	}
 
+	return ts
+}
+
+// LoadTransitionState loads the transition state at root into
+// the "active" transition state.
+func (db *cachingDB) LoadTransitionState(root common.Hash) {
+	ts := db.getTransitionState(root)
+	if ts == nil {
+		panic("nil ts, should not happen")
+	}
+
 	// Copy so that the CurrentAddress pointer in the map
 	// doesn't get overwritten.
 	db.CurrentTransitionState = ts.Copy()
 
 	fmt.Println("loaded transition state", "storage processed", db.CurrentTransitionState.StorageProcessed, "addr", db.CurrentTransitionState.CurrentAccountAddress, "slot hash", db.CurrentTransitionState.CurrentSlotHash, "root", root, "ended", db.CurrentTransitionState.ended, "started", db.CurrentTransitionState.started)
 	debug.PrintStack()
 }
+
+// GetTransitionStateCopy gets the current transition state and
+// makes a copy of it to avoid concurrency issues. It doesn't
+// make it the "active" state.
+func (db *cachingDB) GetTransitionStateCopy(root common.Hash) *TransitionState {
+	ts := db.getTransitionState(root)
+	return ts.Copy()
+}
+
+func (db *cachingDB) GetCurrentTransitionStateCopy() *TransitionState {
+	return db.CurrentTransitionState.Copy()
+}
+
+func (db *cachingDB) SetActiveTransitionState(ts *TransitionState) {
+	db.CurrentTransitionState = ts
```

Comment:
Shouldn't we use a lock here?
Despite you tried making the assignment "more atomic" now (all fields at once), this still can have a race.